File: src/main/java/cz/cvut/fel/plichjan/distmesh/BoundNodes.java
package cz.cvut.fel.plichjan.distmesh;

import cz.cvut.fel.plichjan.distmesh.matlab.BoundBar;
import cz.cvut.fel.plichjan.distmesh.matlab.Matlab;
import cz.cvut.fel.plichjan.distmesh.result.Mesh;
import delaunay.Pnt;

import java.util.ArrayList;
import java.util.List;
import java.util.Set;
import java.util.TreeSet;

public class BoundNodes {
    public List<Pnt> call(Mesh mesh, List<BoundBar> boundBars) {
        return call(mesh.getP(), boundBars);
    }

    public List<Pnt> call(double[][] p, List<BoundBar> boundBars) {
        final Set<Integer> indexes = new TreeSet<Integer>();
        for (BoundBar bar : boundBars) {
            indexes.add(bar.getA());
            indexes.add(bar.getB());
        }

        final ArrayList<Pnt> pnts = new ArrayList<Pnt>(indexes.size());
        for (Integer i : indexes) {
            pnts.add(new Pnt(p[i]));
        }
        return pnts;
    }
}

File: src/main/kotlin/cz/cvut/fel/plichjan/distmesh/BoundNodes.kt
package cz.cvut.fel.plichjan.distmesh

import cz.cvut.fel.plichjan.distmesh.matlab.BoundBar
import cz.cvut.fel.plichjan.distmesh.result.Mesh
import delaunay.Pnt
import java.util.*

class BoundNodes {
    fun call(mesh: Mesh, boundBars: List<BoundBar>): List<Pnt> {
        return call(mesh.p!!, boundBars)
    }

    fun call(p: Array<DoubleArray>, boundBars: List<BoundBar>): List<Pnt> {
        val indexes = TreeSet<Int>()
        for (bar in boundBars) {
            indexes.add(bar.a)
            indexes.add(bar.b)
        }

        val pnts = ArrayList<Pnt>(indexes.size)
        for (i in indexes) {
            pnts.add(Pnt(*p[i]))
        }
        return pnts
    }
}
